src/main/java/com/google/devtools/build/lib/remote/RemoteActionContextProvider.java

```diff
@@ -88,19 +88,20 @@ public void registerActionContexts(ExecutorBuilder executorBuilder) {
     String buildRequestId = env.getBuildRequestId();
     String commandId = env.getCommandId().toString();
 
-    if (executor == null) {
-      RemoteSpawnCache spawnCache =
-          new RemoteSpawnCache(
-              env.getExecRoot(),
-              remoteOptions,
-              cache,
-              buildRequestId,
-              commandId,
-              env.getReporter(),
-              digestUtil,
-              filesToDownload);
+    // Always provide a spawnCache as this allows no-remote-exec actions
+    // to remotely cache actions.
+    RemoteSpawnCache spawnCache =
```

Comment:
It would be even better to avoid using the cache w/ remote execution, and only apply it when the action ends up not being executed remotely.

Reply:
I think the `spawnCache` needs to be provided here as this registration function is only called once when the executor is initialized. I've introduced a check inside of `AbstractSpawnStrategy` to avoid using this cache if we're using a `RemoteSpawnRunner`

Reply:
Oh, I see. The point is that you *want* to use the remote cache for *local actions* even when the remote executor is enabled.

Reply:
Yes, in the case of certain rules being incompatible with remote execution, this allows a gradual move to remote execution by tagging targets which fail remotely as `no-remote-exec`. Currently this is quite a pain because of the relating issue causing no caching for these targets.